data/adapters/categorical_adapter.py

```diff
@@ -0,0 +1,212 @@
+from __future__ import annotations
```

Comment:
It would be really helpful to reduce the code duplication across the different adapter files and classes (for code maintenance, in particular). If you don't make changes on this now, please create a new issue and add TODOs in the corresponding files.